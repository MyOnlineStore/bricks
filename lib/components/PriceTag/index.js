!function(e,r){"object"==typeof exports&&"object"==typeof module?module.exports=r(require("react"),require("styled-components")):"function"==typeof define&&define.amd?define("bricks",["react","styled-components"],r):"object"==typeof exports?exports.bricks=r(require("react"),require("styled-components")):e.bricks=r(e.react,e["styled-components"])}(window,function(e,r){return function(e){var r={};function t(n){if(r[n])return r[n].exports;var o=r[n]={i:n,l:!1,exports:{}};return e[n].call(o.exports,o,o.exports,t),o.l=!0,o.exports}return t.m=e,t.c=r,t.d=function(e,r,n){t.o(e,r)||Object.defineProperty(e,r,{enumerable:!0,get:n})},t.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},t.t=function(e,r){if(1&r&&(e=t(e)),8&r)return e;if(4&r&&"object"==typeof e&&e&&e.__esModule)return e;var n=Object.create(null);if(t.r(n),Object.defineProperty(n,"default",{enumerable:!0,value:e}),2&r&&"string"!=typeof e)for(var o in e)t.d(n,o,function(r){return e[r]}.bind(null,o));return n},t.n=function(e){var r=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(r,"a",r),r},t.o=function(e,r){return Object.prototype.hasOwnProperty.call(e,r)},t.p="",t(t.s=567)}({0:function(r,t){r.exports=e},1:function(e,r,t){"use strict";t.d(r,"b",function(){return i}),t.d(r,"d",function(){return c}),t.d(r,"a",function(){return a});var n=t(4),o=n,u=o.default,i=o.css,c=(o.createGlobalStyle,o.keyframes),a=o.ThemeProvider;r.c=u},4:function(e,t){e.exports=r},567:function(e,r,t){"use strict";t.r(r);var n,o=t(0),u=t.n(o),i=function(e,r){return Object.defineProperty?Object.defineProperty(e,"raw",{value:r}):e.raw=r,e},c=t(1).c.span.withConfig({displayName:"style__StyledPriceTag",componentId:"sc-4a8vmm-0"})(n||(n=i(["\n    ",";\n"],["\n    ",";\n"])),function(e){var r=e.theme;return"\n            position: relative;\n\n            sup {\n                font-size: .7em;\n            }\n\n            "+(!0===e.strikethrough?"&::after {\n                    content: '';\n                    width: 100%;\n                    height: 2px;\n                    left: 0;\n                    top: 50%;\n                    margin-top: -2px;\n                    position: absolute;\n                    background: "+r.PriceTag.strikethroughColor+";\n                    opacity: .7;\n                    transform: rotate(-8deg);\n                }":"")+"\n        "}),a=function(e,r,t){var n=r.fractionFormat,o=r.superScriptFraction;return t&&"none"===n?"":t&&"dash"===n?"-":void 0!==o&&o?u.a.createElement("sup",{key:"fraction"},e):e},f=function(e,r){return!0===r.hideCurrency?"":e},s=function(e,r,t){return"none"===r.fractionFormat&&t?"":e},l=function(e){return"fraction"===e.type&&0===parseInt(e.value,10)},d=function(e){return("integer"===e.type||"fraction"===e.type)&&0!==parseInt(e.value,10)},p=function(e,r){return{isRound:!!l(r)||e.isRound,isFree:!d(r)&&e.isFree}};r.default=function(e){var r=new Intl.NumberFormat(e.locale,{style:"currency",currency:e.currency}).formatToParts(e.value),t=r.reduce(p,{isRound:!1,isFree:!0}),n=r.map(function(r){switch(r.type){case"fraction":return a(r.value,e,t.isRound);case"currency":return f(r.value,e);case"decimal":return s(r.value,e,t.isRound);case"literal":return e.hideCurrency&&!0===e.hideCurrency?void 0:r.value;default:return r.value}});return u.a.createElement(c,{strikethrough:e.strikethrough},t.isFree&&void 0!==e.freeLabel?e.freeLabel:n)}}})});
//# sourceMappingURL=index.js.map