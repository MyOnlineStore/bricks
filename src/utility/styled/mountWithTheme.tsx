import { mount, ReactWrapper, shallow } from 'enzyme';
import React from 'react';
import MosTheme from '../../themes/MosTheme';

/* tslint:disable */
const mountWithTheme = (component:JSX.Element):ReactWrapper => {
    const ThemeProvider = (shallow(<MosTheme />) as any).dive();

    const context = ThemeProvider
        .instance()
        .getChildContext();

    return mount(component, { context, childContextTypes: ThemeProvider.childContextTypes });
};
/* tslint:enable */

export default mountWithTheme;
