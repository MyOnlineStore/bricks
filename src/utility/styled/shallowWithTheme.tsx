import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import MosTheme from '../../themes/MosTheme';

const shallowWithTheme = (component:JSX.Element):ShallowWrapper => {
    /* tslint:disable */
    const context = shallow(<MosTheme />)
        .dive()
        .instance()
        .getChildContext();
    /* tslint:enable */

    return shallow(component, { context });
};

export default shallowWithTheme;
