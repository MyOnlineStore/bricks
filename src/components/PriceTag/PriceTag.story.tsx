import { storiesOf } from '@storybook/react';
import React from 'react';
import PriceTag from '.';

storiesOf('PriceTag', module)
    .add('default', () => (
        <PriceTag
            parts={[
                { type: 'currency' , value: '€' },
                { type: 'literal' , value: ' ' },
                { type: 'integer', value: '10' },
                { type: 'decimal', value: ',' },
                { type: 'fraction', value: '00' },
            ]}
        />
    ))
    .add('With a hidden currency', () => (
        <PriceTag
            parts={[
                { type: 'currency' , value: '€' },
                { type: 'literal' , value: ' ' },
                { type: 'integer', value: '10' },
                { type: 'decimal', value: ',' },
                { type: 'fraction', value: '20' },
            ]}
            hideCurrency
        />
    ))
    .add('With a hidden currency and superscript franction', () => (
        <PriceTag
            parts={[
                { type: 'currency' , value: '€' },
                { type: 'literal' , value: ' ' },
                { type: 'integer', value: '10' },
                { type: 'decimal', value: ',' },
                { type: 'fraction', value: '20' },
            ]}
            superScriptFraction
            hideCurrency
        />
    ))
    .add('With a dashed fraction', () => (
        <PriceTag
            parts={[
                { type: 'currency' , value: '€' },
                { type: 'literal' , value: ' ' },
                { type: 'integer', value: '10' },
                { type: 'decimal', value: ',' },
                { type: 'fraction', value: '00' },
            ]}
            showDash
        />
    ))
    .add('With a hidden fraction on a round value', () => (
        <PriceTag
            parts={[
                { type: 'currency' , value: '€' },
                { type: 'literal' , value: ' ' },
                { type: 'integer', value: '10' },
                { type: 'decimal', value: ',' },
                { type: 'fraction', value: '00' },
            ]}
            hideZeros
        />
    ))
    .add('With a fraction in superscript', () => (
        <PriceTag
            parts={[
                { type: 'currency' , value: '€' },
                { type: 'literal' , value: ' ' },
                { type: 'integer', value: '10' },
                { type: 'decimal', value: ',' },
                { type: 'fraction', value: '20' },
            ]}
            superScriptFraction
        />
    ))
    .add('With showDash and superscript on a round value', () => (
        <PriceTag
            parts={[
                { type: 'currency' , value: '€' },
                { type: 'literal' , value: ' ' },
                { type: 'integer', value: '10' },
                { type: 'decimal', value: ',' },
                { type: 'fraction', value: '00' },
            ]}
            showDash
            superScriptFraction
        />
    ));
