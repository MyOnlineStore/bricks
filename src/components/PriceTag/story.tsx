import { boolean, select, text } from '@storybook/addon-knobs/react';
import { storiesOf } from '@storybook/react';
import React from 'react';
import PriceTag from '.';
import Text from '../Text';

storiesOf('PriceTag', module).add('Default', () => (
    <Text>
        <PriceTag
            value={text('price', '1.00')}
            freeLabel={text('freeLabel', 'free!')}
            currency={select('currency', ['USD', 'EUR', 'JPY', 'GBP', 'AUD'], 'USD')}
            locale={select('locale', ['en-US', 'nl-NL', 'de-DE', 'jp-JP'], 'en-US')}
            fractionFormat={select('fractionFormat', [undefined, 'dash', 'none'], undefined)}
            hideCurrency={boolean('hideCurrency', false)}
            superScriptFraction={boolean('superScriptFraction', false)}
            strikethrough={boolean('strikethrough', false)}
        />
    </Text>
));
