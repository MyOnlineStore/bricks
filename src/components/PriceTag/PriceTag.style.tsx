import { StyledComponentClass as _S } from 'styled-components';
import styled from '../../utility/styled';
import PriceTag from './PriceTag.template';

const StyledPriceTag = styled(PriceTag)`
    color: ${({ theme }):string => theme.priceTag.color};
    font-family: ${({ theme }):string => theme.priceTag.fontFamily};
    font-size: ${({ theme }):string => theme.priceTag.fontSize};
`;

export default StyledPriceTag;
