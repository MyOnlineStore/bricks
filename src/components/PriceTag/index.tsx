import React, { Component } from 'react';
import { StyledType } from '../../utility/styled';
import StyledPriceTag from './style';
import formatFraction from './formatters/formatFraction';
import formatCurrency from './formatters/formatCurrency';

type PropsType = StyledType & {
    hideCurrency?: boolean;
    superScriptFraction?: boolean;
    value: string;
    freeLabel?: string;
    strikethrough?: boolean;
    fractionFormat?: 'none' | 'dash';
    locale: string;
    currency: string;
};

type PartTypeType =
    | 'currency'
    | 'decimal'
    | 'fraction'
    | 'group'
    | 'infinity'
    | 'integer'
    | 'literal'
    | 'minusSign'
    | 'nan'
    | 'plusSign'
    | 'percentSign';

type StateType = {
    price: string;
    fraction: string;
    currency: string;
    currencyAlignment: 'left' | 'right';
    decimalSeperator: string;
};

type PartType = {
    type: PartTypeType;
    value: string;
};

type NumberFormatter = Intl.NumberFormat & {
    formatToParts(value: number): Array<PartType>;
};

type StatsType = {
    isRound: boolean;
    isFree: boolean;
};

class PriceTag extends Component<PropsType, StateType> {
    private formatter: NumberFormatter;

    public constructor(props: PropsType) {
        super(props);
        this.setFormatter(props.locale, props.currency);
        this.state = {
            price: '',
            fraction: '',
            currency: '',
            currencyAlignment: 'left',
            decimalSeperator: '.',
        };
    }

    private isRound = (part: PartType): boolean => part.type === 'fraction' && parseInt(part.value, 10) === 0;

    private isFree = (part: PartType): boolean => {
        return (part.type === 'integer' || part.type === 'fraction') && parseInt(part.value, 10) !== 0;
    };

    private setFormatter(locale: string, currency: string): void {
        this.formatter = new Intl.NumberFormat(locale, { style: 'currency', currency }) as NumberFormatter;
    }

    private parse(value: string): number {
        const stripped = value.replace(new RegExp(`[^0-9${this.state.decimalSeperator}]`, 'g'), '');
        const parsed = parseFloat(stripped.replace(this.state.decimalSeperator, '.'));

        return !isNaN(parsed) ? parseFloat(parsed.toFixed(2)) : 0;
    }

    private deriveStatsFromPart = (initialStats: StatsType, part: PartType): StatsType => ({
        isRound: this.isRound(part) ? true : initialStats.isRound,
        isFree: this.isFree(part) ? false : initialStats.isFree,
    });

    private format(value: string): string {
        this.setState({ fraction: '' });

        return this.formatter
            .formatToParts(this.parse(value))
            .filter((part, index, parts) => {
                const { value } = part;
                const stats = parts.reduce(this.deriveStatsFromPart, { isRound: false });

                switch (part.type) {
                    case 'currency': {
                        const currencyAlignment = index === parts.length - 1 ? 'right' : 'left';
                        if (this.state.currencyAlignment !== currencyAlignment || this.state.currency !== part.value) {
                            this.setState({ currency: value, currencyAlignment });
                        }

                        return false;
                    }
                    case 'decimal': {
                        if (this.props.fractionFormat === 'none' && stats.isRound) return false;

                        this.setState({ decimalSeperator: value });

                        return true;
                    }
                    case 'literal':
                        return false;

                    case 'fraction':
                        this.setState({ fraction: value });

                        return false;
                    default:
                        return true;
                }
            })
            .reduce((combined: string, { value }: PartType): string => `${combined}${value}`, '');
    }

    private renderFreeLabel = (isFree: boolean, label?: string): string => (isFree && label !== undefined ? label : '');

    private renderPrice(isRound: boolean): JSX.Element | string {
        const fraction = formatFraction(this.state.fraction, this.props, isRound);

        const price = (
            <>
                {this.state.price}
                {fraction}
            </>
        );

        const priceElement = formatCurrency(
            this.props.hideCurrency !== undefined ? this.props.hideCurrency : false,
            this.state.currencyAlignment,
            this.state.currency,
            price,
        );

        return priceElement;
    }

    public componentDidUpdate(prevProps: PropsType): void {
        if (prevProps !== this.props) {
            this.setFormatter(this.props.locale, this.props.currency);
            this.setState({ price: this.format(this.props.value) });
        }
    }

    public render(): JSX.Element {
        const formatter = new Intl.NumberFormat(this.props.locale, {
            style: 'currency',
            currency: this.props.currency,
        }) as NumberFormatter;

        const parts = formatter.formatToParts(this.parse(this.props.value));

        const stats = parts.reduce(this.deriveStatsFromPart, {
            isRound: false,
            isFree: true,
        });

        return (
            <StyledPriceTag strikethrough={this.props.strikethrough}>
                {this.renderFreeLabel(stats.isFree, this.props.freeLabel) || this.renderPrice(stats.isRound)}
            </StyledPriceTag>
        );
    }
}

export default PriceTag;
export { PropsType, PartType, StateType };
