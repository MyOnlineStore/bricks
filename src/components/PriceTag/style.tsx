import styled from '../../utility/styled';
import ThemeTools from '../../themes/ExperimentalCustomTheme/ThemeTools';

type PriceTagThemeType = {
    strikethroughColor: string;
};

type PropsType = {
    strikethrough?: boolean;
    strikethroughStyle?: 'diagonal' | 'horizontal';
};

const StyledPriceTag = styled.span<PropsType>`
    ${({ theme, strikethrough, strikethroughStyle }): string => {
        return `
            position: relative;

            sup {
                font-size: .7em;
            }

            ${
                strikethrough === true
                    ? `&::after {
                    content: '';
                    width: 100%;
                    height: 2px;
                    left: 0;
                    top: 50%;
                    margin-top: ${strikethroughStyle === 'horizontal' ? '-1px' : '-2px'};
                    position: absolute;
                    background: ${theme.PriceTag.strikethroughColor};
                    opacity: .7;
                    transform: rotate(${strikethroughStyle === 'horizontal' ? '0deg' : '-8deg'});
                }`
                    : ''
            }
        `;
    }};
`;

const composePriceTagTheme = (themeTools: ThemeTools): PriceTagThemeType => {
    const { colors } = themeTools.themeSettings;

    return {
        strikethroughColor: colors.grey.lighter2,
    };
};

export default StyledPriceTag;
export { PriceTagThemeType, composePriceTagTheme };
