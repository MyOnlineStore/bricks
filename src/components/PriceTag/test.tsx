import toJson from 'enzyme-to-json';
import React from 'react';
import PriceTag from '.';
import { mountWithTheme } from '../../utility/styled/testing';

describe('PriceTag', () => {
    it('should render with a hidden currency', () => {
        const component = mountWithTheme(<PriceTag locale="nl-NL" currency="EUR" value="10,20" hideCurrency />).find(
            PriceTag,
        );

        expect(component.text()).not.toContain('€');
    });

    it('should render a base price', () => {
        const component = mountWithTheme(<PriceTag locale="de-DE" currency="EUR" value="10,20" />);

        component.setProps({ value: '10,45' });
        component.update();

        expect(component.text()).toContain('10,45');
    });

    it('should not break when an incorrect price is provided', () => {
        const component = mountWithTheme(<PriceTag locale="de-DE" currency="EUR" value="🐛" />).find(PriceTag);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('should render an action price', () => {
        const component = mountWithTheme(<PriceTag locale="nl-NL" currency="EUR" value="10,20" strikethrough />).find(
            PriceTag,
        );

        expect(toJson(component)).toMatchSnapshot();
    });

    it('should render with a hidden currency and superscript franction', () => {
        const component = mountWithTheme(
            <PriceTag locale="nl-NL" currency="EUR" value="10,20" superScriptFraction hideCurrency />,
        ).find(PriceTag);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('should render with a dashed fraction', () => {
        const component = mountWithTheme(
            <PriceTag locale="nl-NL" currency="EUR" value="10.00" fractionFormat="dash" />,
        ).find(PriceTag);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('should render with a hidden fraction on a round value', () => {
        const component = mountWithTheme(
            <PriceTag locale="nl-NL" currency="EUR" value="10.00" fractionFormat="none" />,
        ).find(PriceTag);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('should render with a fraction in superscript', () => {
        const component = mountWithTheme(
            <PriceTag locale="nl-NL" currency="EUR" value="10,20" superScriptFraction />,
        ).find(PriceTag);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('should render with showDash and superscript on a round value', () => {
        const component = mountWithTheme(
            <PriceTag locale="nl-NL" currency="EUR" value="0,00" fractionFormat="dash" superScriptFraction />,
        ).find(PriceTag);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('should render a label with the price is 0', () => {
        const component = mountWithTheme(<PriceTag locale="nl-NL" currency="EUR" value="0" freeLabel="free" />).find(
            PriceTag,
        );

        expect(toJson(component)).toMatchSnapshot();
    });
});
