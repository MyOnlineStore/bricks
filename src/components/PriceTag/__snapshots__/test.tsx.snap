// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceTag should not break when an incorrect price is provided 1`] = `
.c0 {
  position: relative;
}

.c0 sup {
  font-size: .7em;
}

<PriceTag
  currency="EUR"
  locale="de-DE"
  value="🐛"
>
  <styled.span>
    <span
      className="c0"
    >
       
    </span>
  </styled.span>
</PriceTag>
`;

exports[`PriceTag should render a label with the price is 0 1`] = `
.c0 {
  position: relative;
}

.c0 sup {
  font-size: .7em;
}

<PriceTag
  currency="EUR"
  freeLabel="free"
  locale="nl-NL"
  value="0"
>
  <styled.span>
    <span
      className="c0"
    >
      free
    </span>
  </styled.span>
</PriceTag>
`;

exports[`PriceTag should render an action price 1`] = `
.c0 {
  position: relative;
}

.c0 sup {
  font-size: .7em;
}

.c0::after {
  content: '';
  width: 100%;
  height: 2px;
  left: 0;
  top: 50%;
  margin-top: -2px;
  position: absolute;
  background: #88979d;
  opacity: .7;
  -webkit-transform: rotate(-8deg);
  -ms-transform: rotate(-8deg);
  transform: rotate(-8deg);
}

<PriceTag
  currency="EUR"
  locale="nl-NL"
  strikethrough={true}
  value="10,20"
>
  <styled.span
    strikethrough={true}
  >
    <span
      className="c0"
    >
       
    </span>
  </styled.span>
</PriceTag>
`;

exports[`PriceTag should render with a dashed fraction 1`] = `
.c0 {
  position: relative;
}

.c0 sup {
  font-size: .7em;
}

<PriceTag
  currency="EUR"
  fractionFormat="dash"
  locale="nl-NL"
  value="10.00"
>
  <styled.span>
    <span
      className="c0"
    >
       
      -
    </span>
  </styled.span>
</PriceTag>
`;

exports[`PriceTag should render with a fraction in superscript 1`] = `
.c0 {
  position: relative;
}

.c0 sup {
  font-size: .7em;
}

<PriceTag
  currency="EUR"
  locale="nl-NL"
  superScriptFraction={true}
  value="10,20"
>
  <styled.span>
    <span
      className="c0"
    >
       
      <sup
        key="fraction"
      />
    </span>
  </styled.span>
</PriceTag>
`;

exports[`PriceTag should render with a hidden currency and superscript franction 1`] = `
.c0 {
  position: relative;
}

.c0 sup {
  font-size: .7em;
}

<PriceTag
  currency="EUR"
  hideCurrency={true}
  locale="nl-NL"
  superScriptFraction={true}
  value="10,20"
>
  <styled.span>
    <span
      className="c0"
    >
      <sup
        key="fraction"
      />
    </span>
  </styled.span>
</PriceTag>
`;

exports[`PriceTag should render with a hidden fraction on a round value 1`] = `
.c0 {
  position: relative;
}

.c0 sup {
  font-size: .7em;
}

<PriceTag
  currency="EUR"
  fractionFormat="none"
  locale="nl-NL"
  value="10.00"
>
  <styled.span>
    <span
      className="c0"
    >
       
    </span>
  </styled.span>
</PriceTag>
`;

exports[`PriceTag should render with showDash and superscript on a round value 1`] = `
.c0 {
  position: relative;
}

.c0 sup {
  font-size: .7em;
}

<PriceTag
  currency="EUR"
  fractionFormat="dash"
  locale="nl-NL"
  superScriptFraction={true}
  value="0,00"
>
  <styled.span>
    <span
      className="c0"
    >
       
      -
    </span>
  </styled.span>
</PriceTag>
`;
